src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java

```diff
@@ -1255,12 +1253,47 @@ private void validateStaticCodeAnalysisSettings(ProgrammingExercise programmingE
+                .requireNonNullElse(programmingExercise.getAuxiliaryRepositories(), new ArrayList<AuxiliaryRepository>()).stream().filter(repo -> repo.getId() != null).toList());
         for (AuxiliaryRepository repo : newAuxiliaryRepositories) {
-            validateAuxiliaryRepository(repo, auxiliaryRepositories);
+            validateAuxiliaryRepository(repo, auxiliaryRepositories, true);
+            auxiliaryRepositories.add(repo);
+        }
+        programmingExercise.setAuxiliaryRepositories(new ArrayList<>());
+        auxiliaryRepositories.forEach(programmingExercise::addAuxiliaryRepository);
+    }
+
+    private void handleAuxiliaryRepositoriesWhenUpdatingExercises(ProgrammingExercise programmingExercise, ProgrammingExercise updatedExercise) {
+        // Get all new (ID is still null) and changed (some string value is changed) auxiliary repositories
+        List<AuxiliaryRepository> newOrEditedAuxiliaryRepositories = new ArrayList<>(updatedExercise.getAuxiliaryRepositories().stream().filter(repo -> {
+            if (repo.getId() == null) {
+                return true;
+            }
+            AuxiliaryRepository auxiliaryRepositoryBeforeUpdate = auxiliaryRepositoryRepository.findById(repo.getId())
+                    .orElseThrow(() -> new IllegalStateException("Edited an existing repository that is not in the data base!"));
+            return !repo.containsSameStringValues(auxiliaryRepositoryBeforeUpdate);
+        }).toList());
+        validateAndUpdateExistingAuxiliaryRepositoriesOfProgrammingExercise(programmingExercise, newOrEditedAuxiliaryRepositories, updatedExercise);
+
+        List<AuxiliaryRepository> removedAuxiliaryRepositories = programmingExercise.getAuxiliaryRepositories().stream()
+                .filter(repo -> updatedExercise.getAuxiliaryRepositories().stream().noneMatch(updatedRepo -> repo.getId().equals(updatedRepo.getId()))).toList();
+
+        auxiliaryRepositoryRepository.deleteAll(removedAuxiliaryRepositories);
+    }
+
+    private void validateAndUpdateExistingAuxiliaryRepositoriesOfProgrammingExercise(ProgrammingExercise programmingExercise,
+            List<AuxiliaryRepository> updatedAuxiliaryRepositories, ProgrammingExercise updatedExercise) {
+        // Get all repositories that are unchanged and are still present in the updated exercise
+        List<AuxiliaryRepository> auxiliaryRepositories = new ArrayList<>(
+                Objects.requireNonNullElse(programmingExercise.getAuxiliaryRepositories(), new ArrayList<AuxiliaryRepository>()).stream()
+                        .filter(existingRepo -> updatedAuxiliaryRepositories.stream().noneMatch((updatedRepo -> existingRepo.getId().equals(updatedRepo.getId())))
+                                && updatedExercise.getAuxiliaryRepositories().stream().anyMatch(updatedRepo -> existingRepo.getId().equals(updatedRepo.getId())))
+                        .toList());
+
+        for (AuxiliaryRepository repo : updatedAuxiliaryRepositories) {
```

Comment:
The end of the method from here is technically a duplication from validateAndAddAuxiliaryRepositoriesOfProgrammingExercise, although I am not sure if it actually increases the readability if an extra layer of helper methods is added

Reply:
Since the method is quite short, I'd say that adding another layer is more confusing than this small duplication